linera-faucet/server/src/lib.rs

```diff
@@ -182,72 +215,282 @@ where
+
+            if let Err(err) = self.execute_batch(batch_requests).await {
+                tracing::error!("Failed to execute batch: {}", err);
+                return Err(err);
+            }
+        }
+    }
+
+    /// Executes a batch of chain creation requests
+    async fn execute_batch(
+        &mut self,
+        batch_requests: Vec<PendingRequest>,
+    ) -> Result<(), anyhow::Error> {
+        // Pre-validate: check rate limiting and existing chains
+        let mut valid_requests = Vec::new();
+
+        for request in batch_requests {
+            // Check if this owner already has a chain
+            {
+                let storage = self.faucet_storage.lock().await;
+                if let Some(existing_description) = storage.get_chain(&request.owner) {
+                    let _ = request.responder.send(Ok(existing_description.clone()));
+                    continue;
+                }
+            }
+
+            valid_requests.push(request);
+        }
+
+        if valid_requests.is_empty() {
+            return Ok(());
+        }
 
-        if self.start_timestamp < self.end_timestamp {
-            let local_time = client.storage_client().clock().current_time();
-            if local_time < self.end_timestamp {
+        // Rate limiting check for the batch
+        if self.config.start_timestamp < self.config.end_timestamp {
+            let local_time = self.client.storage_client().clock().current_time();
+            if local_time < self.config.end_timestamp {
                 let full_duration = self
+                    .config
                     .end_timestamp
-                    .delta_since(self.start_timestamp)
+                    .delta_since(self.config.start_timestamp)
                     .as_micros();
-                let remaining_duration = self.end_timestamp.delta_since(local_time).as_micros();
-                let balance = client.local_balance().await?;
-                let Ok(remaining_balance) = balance.try_sub(self.amount) else {
-                    return Err(Error::new("The faucet is empty."));
+                let remaining_duration = self
+                    .config
+                    .end_timestamp
+                    .delta_since(local_time)
+                    .as_micros();
+                let balance = self.client.local_balance().await?;
+
+                let total_amount = self
+                    .config
+                    .amount
+                    .saturating_mul(valid_requests.len() as u128);
+                let Ok(remaining_balance) = balance.try_sub(total_amount) else {
+                    // Not enough balance - reject all requests
+                    for request in valid_requests {
+                        let _ = request
+                            .responder
+                            .send(Err(Error::new("The faucet is empty.")));
+                    }
+                    return Ok(());
```

Comment:
Shouldn't we try to handle as many requests as possible rather than failing the whole batch? Maybe we can handle `n-1` requests and the last one fails the batch.